src/vs/workbench/api/browser/mainThreadLanguageFeatures.ts

```diff
@@ -295,52 +295,28 @@ export class MainThreadLanguageFeatures extends Disposable implements MainThread
-		this._registrations.set(handle, combinedDisposable(
-			this._languageFeaturesService.documentHighlightProvider.register(selector, <languages.DocumentHighlightProvider>{
-				provideDocumentHighlights: (model: ITextModel, position: EditorPosition, token: CancellationToken): Promise<languages.DocumentHighlight[] | undefined> => {
-					return this._proxy.$provideDocumentHighlights(handle, model.uri, position, token);
-				}
-			}),
-			this._languageFeaturesService.multiDocumentHighlightProvider.register(selector, <languages.MultiDocumentHighlightProvider>{
-				provideMultiDocumentHighlights: (model: ITextModel, position: EditorPosition, otherModels: ITextModel[], token: CancellationToken): Promise<Map<URI, languages.DocumentHighlight[]>> => {
-					const word = model.getWordAtPosition({
-						lineNumber: position.lineNumber,
-						column: position.column
-					});
-
-					const res = this._proxy.$provideDocumentHighlights(handle, model.uri, position, token);
-					return res.then(data => {
-						const result = new Map<URI, languages.DocumentHighlight[]>();
-						if (isFalsyOrEmpty(data) || !data) {
-							return result;
-						}
-						result.set(model.uri, data);
-
-						if (!word) {
-							return result;
-						}
-
-						for (const otherModel of otherModels) {
-							if (otherModel.isDisposed()) {
-								continue;
-							}
-
-							const matches = otherModel.findMatches(word.word, true, false, true, USUAL_WORD_SEPARATORS, false);
-							const highlights = matches.map(m => ({
-								range: m.range,
-								kind: languages.DocumentHighlightKind.Text
-							}));
-
-							if (highlights) {
-								result.set(otherModel.uri, highlights);
-							}
-						}
+		this._registrations.set(handle, this._languageFeaturesService.documentHighlightProvider.register(selector, <languages.DocumentHighlightProvider>{
+			provideDocumentHighlights: (model: ITextModel, position: EditorPosition, token: CancellationToken): Promise<languages.DocumentHighlight[] | undefined> => {
+				return this._proxy.$provideDocumentHighlights(handle, model.uri, position, token);
+			}
+		}));
+	}
 
-						return result;
+	$registerMultiDocumentHighlightProvider(handle: number, selector: IDocumentFilterDto[]): void {
+		this._registrations.set(handle, this._languageFeaturesService.multiDocumentHighlightProvider.register(selector, <languages.MultiDocumentHighlightProvider>{
+			provideMultiDocumentHighlights: (model: ITextModel, position: EditorPosition, otherModels: ITextModel[], token: CancellationToken): Promise<Map<URI, languages.DocumentHighlight[]> | undefined> => {
+				return this._proxy.$provideMultiDocumentHighlights(handle, model.uri, position, otherModels.map(model => model.uri), token).then(dto => {
+					if (isFalsyOrEmpty(dto) || !dto) {
+						return undefined;
+					}
+					const result = new ResourceMap<languages.DocumentHighlight[]>();
+					dto.forEach(value => {
+						result.set(URI.revive(value.uri), value.highlights);
```

Comment:
This will lose highlights when having "mixed" results, like `[uri1, A], [uri2, B], [uri1, C]`. You will now nuke `A` instead of combining it with `C`